### ISSUE TITLE
Skip not null properties on target

### ISSUE THREAD

**Issue by thiagogjt** [mapstruct/mapstruct#1443]
Hi folks!
Is there a way to fill properties on target only if it have null values?
Ex:

Source
prop1 = "val1"
prop2 = "val2"

Target
prop1 = null
prop2 = "val3"

Result
prop1 = "val1"
prop2 = "val3"

in other words:
```
if (target.prop1 == null) {
   target.setProp1(source.prop1);
}
if (target.prop2 == null) {
   target.setProp2(source.prop2);
}

```

**Comment by filiphr** [comment]
Please use [Stackoverflow](https://stackoverflow.com/questions/tagged/mapstruct) or [gitter](https://gitter.im/mapstruct/mapstruct-users) channel for questions.

Have you read [Controlling checking result for 'null' properties in bean mapping
](http://mapstruct.org/documentation/stable/reference/html/#checking-source-property-for-null-arguments. You would need to set it to [`NullValueCheckStrategy.ALWAYS`](http://mapstruct.org/documentation/stable/api/org/mapstruct/NullValueCheckStrategy.html#ALWAYS)

**Comment by thiagogjt** [comment]
Hi filiphr!

Thanks for the response, i checked on the documentation but i want to perform a null check on the target before set the property not on source.
I have sent the question on gitter.

Again thanks for your time.
Brest regards

**Comment by filiphr** [comment]
@thiagogjt sorry it's my mistake. I understood it wrong. This is currently not possible with MapStruct. The `null` of a target is only checked when we have nested target mappings, but in this case only a new object is set.

**Comment by filiphr** [comment]
@thiagogjt there might be a way to achieve this if you update the properties that are incoming, that way the properties from the higher priority bean would override the ones in the incoming and those that are not present in the higher priority would stay in the incoming.

You could also do this by a small combination of mappings. You would first map the incoming into a result, and then update the result with the properties from the higher priority one. Although that might not work due to #879 (you'll have to try).

>I have thought about and if we have the possibility to assign an expression in mapping to filter the assignment we can achieve others use cases as well.

What do you mean to have the possibility to assign an expression?

**Comment by thiagogjt** [comment]
My situation is quite similar to #879.

There is an example of my idea:

```
@Mapping(target = "foo",
         filterExpression = "java( target.getFoo() != null  )"
)
```
The result will be something like

```
if([filterExpression]) {
   target.setXXX(souce.getXXX);
}
```

In that whay other situations can be acomplished like overide only if target(or source) has certain values or pass in some validation.
Let me now what you think about it.

Best regards.

**Comment by artem-emelin** [comment]
For me it would be also a nice option to have for all fields in the object.
We have streaming updates coming in but we need to update fields only if target fields is `null`

**Comment by c0ff75** [comment]
It's very useful feature. It needed when update must doing without overwrite fields in target object.
It's needed in very much cases.

**Comment by filiphr** [comment]
The core mapstruct team is currently not focused on this feature as we believe it is a corner case. We have recently added support for conditional mappings, so something like this could be achieved with that. Have a look at https://mapstruct.org/news/2021-07-18-mapstruct-1_5_0_Beta1-is-out/#conditional-mapping

**Comment by thejeff77** [comment]
@filiphr - I'd argue this might be more common than you'd think. 

When you have objects that reference each other - you get an infinite loop pretty quick. The only way you can fix this currently is by configuring one of the mappers to set one of the references to null every time - which is a mapping that isn't complete.

There is no way at all to successfully map a dual-reference object unless you have a break condition checking for not null on the target object.

**Comment by mcanalesmayo** [comment]
> My data came from different sources and this sources have a priority order. In one of that cases the data came from a local sevice and have his properties enhanced by a remote one, but only the properties that are missing should be filled. I have thought about and if we have the possibility to assign an expression in mapping to filter the assigngment we can achieve others use cases as well.

We have the same requirement as @thiagogjt, this feature would definitely be useful for us as well. As workaround, we are having an `@AfterMapping` callback to set all these properties if target property is not null, but it's really cumbersome as this is all boilerplate code.

**Comment by codermuyilee** [comment]
mark

**Comment by smaugfm** [comment]
Yes, something like `NullTargetValuePropertyMappingStrategy` would be great.

**Comment by bduisenov** [comment]
This can be achieved using `conditionExpression` as follows:
```java
    @Mapping(target = "someId", source = "someId", conditionExpression = "java(someTargetObject.getSomeId() == null)")
```
Here, `someTargetObject` is an autogenerated name by mapstruct, and you can find it in the generated implementation code.
